src/main/java/com/epam/aidial/core/util/BufferingReadStream.java

```diff
@@ -109,9 +118,49 @@ public synchronized ReadStream<Buffer> endHandler(Handler<Void> handler) {
         return this;
     }
 
+    public synchronized void end(HttpServerResponse response) {
+        if (lastChunkPos != -1) {
+            Buffer lastChunk = content.slice(lastChunkPos, content.length());
+            response.end(lastChunk);
+        } else {
+            response.end();
+        }
+    }
+
     private synchronized void handleChunk(Buffer chunk) {
+        int pos = content.length();
         content.appendBuffer(chunk);
-        notifyOnChunk(chunk);
+        if (lastChunkPos != -1) {
+            // stop streaming
+            return;
+        }
+        if (isStreaming && isLastChunk(content)) {
+            lastChunkPos = pos;
+        } else {
+            notifyOnChunk(chunk);
+        }
+    }
+
+    private static boolean isLastChunk(Buffer content) {
+        int i = skipWhitespaces(content, content.length() - 1);
+        String lastMessage = "data: [DONE]";
+        int j = lastMessage.length() - 1;
+        for (; i >= 0 && j >= 0; i--, j--) {
```

Comment:
Gives true for partially matched content: e.g. ], ONE]

Reply:
Check that the content has enough bytes to match

Reply:
fixed